src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/TagHelperCollection.Builder.Enumerator.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Razor.Language;
+
+public abstract partial class TagHelperCollection
+{
+    public sealed partial class Builder
+    {
+        public ref struct Enumerator(Builder builder)
+        {
+            private int _index = -1;
+
+            public readonly TagHelperDescriptor Current => builder[_index];
```

Comment:
>builder[_index]; [](http://example.com/codeflow?start=59&length=16)

Do you want to check for invalid usage like TagHelperCollection.Enumerator does?

Reply:
No, and I'd prefer not to do it in `TagHelperCollection.Enumerator.Current` either. :smile:

`RefBuilder.Enumerator.Current`, `Builder.Enumerator.Current`, and `Builder.EnumeratorImpl.Current` will all fail if called before enumeration begins because they're each indexing into a builder with -1. They don't need extra checks.

`TagHelperCollection.Enumerator.Current` needs an extra check because it's much more complicated and tracks the current value in a separate field. Without the extra check `TagHelperCollection.Enuemrator.Current` would just return `null` if called before `MoveNext`.